core/shared/src/main/scala/bobcats/Hotp.scala

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright 2021 Typelevel
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package bobcats
+
+import cats.effect.kernel.Sync
+import cats.syntax.functor._
+import scodec.bits.ByteVector
+
+sealed trait Hotp[F[_]] {
+  def generate(
+      key: SecretKey[HmacAlgorithm.SHA1.type],
+      movingFactor: Long,
+      digits: Int
+  ): F[Int]
+
+  def generate(
+      key: SecretKey[HmacAlgorithm.SHA1.type],
+      movingFactor: Long
+  ): F[Int] =
+    generate(key, movingFactor, digits = 6)
+}
+
+private[bobcats] trait UnsealedHotp[F[_]] extends Hotp[F]
+
+object Hotp {
+  private val powersOfTen =
+    Array(1, 10, 100, 1000, 10000, 100000, 1000000, 10000000, 100000000, 1000000000)
+
+  def apply[F[_]](implicit hotp: Hotp[F]): hotp.type = hotp
+
+  implicit def forSync[F[_]](implicit F: Sync[F], H: Hmac[F]): Hotp[F] =
+    new UnsealedHotp[F] {
+      override def generate(
+          key: SecretKey[HmacAlgorithm.SHA1.type],
+          movingFactor: Long,
+          digits: Int
+      ): F[Int] = {
+        require(digits >= 6, s"digits must be at least 6, was $digits")
+        require(digits < 10, s"digits must be less than 10, was $digits")
```

Comment:
Let's wrap this in `F.catchNonFatal` so that errors suspend properly into `F`.

Reply:
Actually, why do we need `Sync` here? Would `Functor` be enough? In which case you can leave the `require` as-is.

Reply:
I think there's a design question here whether we need an `Hotp` capability trait actually since it's not hiding bigger constraints. We could implement these as methods on an object, that take `Functor` and `Hmac` constraints. I'd need to look to other Typelevel libs for inspiration on the right way to do this :)

Reply:
Ah, that's interesting :) We have `ApplicativeThrow#catchNonFatal` and there may be similar methods in there.